internal/proxy/task_delete.go

```diff
@@ -304,164 +392,219 @@ func (dt *deleteTask) getStreamingQueryAndDelteFunc(stream msgstream.MsgStream,
-				return err
-			}
+		// query or produce task failed
+		if dr.err != nil {
+			return dr.err
+		}
+		return nil
+	}
+}
 
-			err = dt.produce(ctx, stream, result.GetIds())
-			if err != nil {
-				log.Warn("query stream for delete produce result failed", zap.Int64("msgID", dt.msgID), zap.Error(err))
-				return err
+func (dr *deleteRunner) receiveQueryResult(ctx context.Context, client querypb.QueryNode_QueryStreamClient, taskCh chan *deleteTask) {
+	defer func() {
+		close(taskCh)
+	}()
+
+	for {
+		result, err := client.Recv()
+		if err != nil {
+			if err == io.EOF {
+				log.Debug("query stream for delete finished", zap.Int64("msgID", dr.msgID))
+				return
 			}
+			dr.err = err
+			return
+		}
+
+		err = merr.Error(result.GetStatus())
+		if err != nil {
+			dr.err = err
+			log.Warn("query stream for delete get error status", zap.Int64("msgID", dr.msgID), zap.Error(err))
+			return
+		}
+
+		task, err := dr.produce(ctx, result.GetIds())
+		if err != nil {
+			dr.err = err
+			log.Warn("produce delete task failed", zap.Error(err))
+			return
 		}
+
+		taskCh <- task
 	}
 }
 
-func (dt *deleteTask) complexDelete(ctx context.Context, plan *planpb.PlanNode, stream msgstream.MsgStream) error {
-	err := dt.lb.Execute(ctx, CollectionWorkLoad{
-		db:             dt.req.GetDbName(),
-		collectionName: dt.req.GetCollectionName(),
-		collectionID:   dt.collectionID,
+func (dr *deleteRunner) complexDelete(ctx context.Context, plan *planpb.PlanNode) error {
+	rc := timerecord.NewTimeRecorder("QueryStreamDelete")
+	var err error
+
+	dr.msgID, err = dr.idAllocator.AllocOne()
+	if err != nil {
+		return err
+	}
+
+	dr.ts, err = dr.tsoAllocatorIns.AllocOne(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = dr.lb.Execute(ctx, CollectionWorkLoad{
+		db:             dr.req.GetDbName(),
+		collectionName: dr.req.GetCollectionName(),
+		collectionID:   dr.collectionID,
 		nq:             1,
-		exec:           dt.getStreamingQueryAndDelteFunc(stream, plan),
+		exec:           dr.getStreamingQueryAndDelteFunc(plan),
 	})
+	dr.result.DeleteCnt = dr.count.Load()
 	if err != nil {
-		log.Warn("fail to get or create dml stream", zap.Error(err))
+		log.Warn("fail to execute complex delete",
+			zap.Int64("deleteCnt", dr.result.GetDeleteCnt()),
+			zap.Duration("interval", rc.ElapseSpan()),
+			zap.Error(err))
 		return err
 	}
 
+	log.Info("complex delete finished", zap.Int64("deleteCnt", dr.result.GetDeleteCnt()), zap.Duration("interval", rc.ElapseSpan()))
 	return nil
 }
 
-func (dt *deleteTask) simpleDelete(ctx context.Context, termExp *planpb.Expr_TermExpr, stream msgstream.MsgStream) error {
-	primaryKeys, numRow, err := getPrimaryKeysFromExpr(dt.schema, termExp)
-	if err != nil {
-		log.Info("Failed to get primary keys from expr", zap.Error(err))
-		return err
-	}
+func (dr *deleteRunner) simpleDelete(ctx context.Context, pk *schemapb.IDs, numRow int64) error {
 	log.Debug("get primary keys from expr",
```

Comment:
should we allocate ID and ts ?

Reply:
> should we allocate ID and ts ?

Runner allocate MsgID and Ts just for query, Dlelete MsgID and Ts will alloc when delete task run. So simple delete should not alloc ID and Ts because has no query.